### ISSUE TITLE
Fixed-width integer types vs built-in types

### ISSUE THREAD

**Issue by oleksandrkozlov** [isocpp/CppCoreGuidelines#1507]
Could we add a guideline about when should we prefer fixed-width integer types (e.g., `int32_t`)  over built-in types (e.g., `int`)?

**Comment by jwakely** [comment]
If you need a specific width type, use one, otherwise don't? Does that really need further detail? Anything more is going to be highly application specific.

**Comment by oleksandrkozlov** [comment]
For me, as a developer, it;s not obvious, that's why I would have detail.

Let me rephrase. You are saying
- use built-in types by default;
- use fixed-width types if a use case requires.

Should we take performance into account (AFAIK built-in types are more efficient)?
Should we take into account portability?

**Comment by jwakely** [comment]
Yes, you use the fixed-width types if you need a fixed width that will have a fixed number of bits on all platforms. It's that simple.

The fixed-width types are just typedefs for the built-in types, so they can't be more efficient.

**Comment by oleksandrkozlov** [comment]
Thank you for the explanation. If you think that this guide is redundant, I guess, the issue may be closed.

**Comment by mguid65** [comment]
I would add that you should use the fixed width types if, obviously, you expect the number of bits to be fixed.

It seems that some developers like to assume that for example, int is 32 bit and long is 64 bit, but of course we know otherwise.

some linters like to suggest to use the fixed width types; for example, this warning from cpplint:

    Use int16/int64/etc, rather than the C type long [runtime/int]

**Comment by ChrisIdema** [comment]
As an embedded firmware engineer I rarely use the build-in types. The fixed width types helps you to set the range and signed-ness of your input arguments and makes you think about edge and corner cases of your inputs. It is also portable. On some microcontrollers an int is 16 bit wide which can lead to bugs if you simply assume that int is 32 bit. 
It is a wrong assumption smaller types are always more efficient. A ordered and packed struct will be smaller, but functions will not necessarily be faster. So you can take it too far.

**Comment by gjones1** [comment]
I'd also add that the if the fixed width type is not supported by the target micro-controller it will not be defined.  This makes porting code from one device to another far more predictable as it's an error if the data type is not supported.

**POI**
This was useful when porting code that assumed an 8-bit char from a micro-controller (8051) to a DSP (ADSP-2717) that only supported 16-bit types only (char, short, int, long etc ... were all 16-bit).  This was a long time ago (hence 8051 and 2171).

**Comment by hsutter** [comment]
Editors' call: Jonathan summarized it well. Also the fixed-width aliases are not required to be available.
